File: scripts/ima-compile.js
import axios from 'axios';
import * as fs from 'fs';
import * as path from 'path';

const IMA_API_KEY = 'rSAWpVuJjTYWqwZSXrNWoHZSl+pvxL+tfPOAiW0nwVEAIQv9oToEMRAXHk+WDzGuyaSTh8CI4A==';
const IMA_CLIENT_ID = '673604a6665155cd973af671ab115321';
const RAW_KB_ID = '2Oor28yUr2eXyEFPBL0bwQU0DsrUEWPW-u9RF9CIUTw=';
const WIKI_KB_ID = 'TcygOiKth0s91W35XJOu4NTAatl-57Bi2g19Wl3jvFw=';

const BASE_URL = 'https://ima.qq.com/openapi';
const HEADERS = {
  'ima-openapi-clientid': IMA_CLIENT_ID,
  'ima-openapi-apikey': IMA_API_KEY,
  'Content-Type': 'application/json'
};

const TOPICS = [
  'Claude Code与Harness Engineering',
  'AI Agent架构',
  'DeepSeek技术',
  'Prompt Engineering与AI工作流',
  'Embedding模型选型',
  'AI科研自动化',
  '个人效能与方法论',
  '知识库构建'
];

const TOPIC_KEYWORDS = {
  'Claude Code与Harness Engineering': ['claude', 'code', 'harness', 'engineering', 'ide', 'trae', 'cursor'],
  'AI Agent架构': ['agent', 'agentic', 'architecture', 'workflow', 'autonomous', 'multi-agent'],
  'DeepSeek技术': ['deepseek', 'moonshot', 'model', 'llm', 'inference'],
  'Prompt Engineering与AI工作流': ['prompt', 'engineering', 'workflow', 'chain', 'rag', 'fine-tune'],
  'Embedding模型选型': ['embedding', 'vector', 'model', 'similarity', 'retrieval'],
  'AI科研自动化': ['research', 'paper', 'arxiv', 'scientific', 'automation'],
  '个人效能与方法论': ['productivity', 'efficiency', 'methodology', 'workflow', 'system'],
  '知识库构建': ['knowledge', 'base', 'kb', 'wiki', 'notes', 'organize']
};

async function request(method, url, data = {}) {
  try {
    const response = await axios({
      method,
      url: `${BASE_URL}${url}`,
      headers: HEADERS,
      data: JSON.stringify(data),
      proxy: false
    });
    return response.data;
  } catch (error) {
    console.error(`API请求失败 ${url}:`, error.response?.data || error.message);
    throw error;
  }
}

async function getWikiKnowledgeList(kbId) {
  const allItems = [];
  let cursor = '';
  let isEnd = false;

  while (!isEnd) {
    const result = await request('POST', '/wiki/v1/get_knowledge_list', {
      knowledge_base_id: kbId,
      cursor,
      limit: 50
    });

    if (result.data?.knowledge_list) {
      allItems.push(...result.data.knowledge_list);
    }
    isEnd = result.data?.is_end || false;
    cursor = result.data?.cursor || '';
  }

  return allItems;
}

async function getNoteContent(noteId) {
  const result = await request('POST', '/note/v1/search_note', {
    search_type: 0,
    query_info: { title: '' },
    start: 0,
    end: 1
  });
  return null;
}

async function fetchUrlContent(url) {
  try {
    const response = await axios.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
      }
    });
    return response.data;
  } catch (error) {
    console.error(`URL抓取失败 ${url}:`, error.message);
    return null;
  }
}

function classifyTopic(title, content) {
  const text = `${title} ${content || ''}`.toLowerCase();
  for (const [topic, keywords] of Object.entries(TOPIC_KEYWORDS)) {
    if (keywords.some(kw => text.includes(kw.toLowerCase()))) {
      return topic;
    }
  }
  return '其他';
}

function generateSummary(content) {
  if (!content || content.length < 50) return '';
  const sentences = content.replace(/[\r\n]+/g, ' ').split(/[。！？.!?]/);
  const summary = sentences.slice(0, 3).filter(s => s.trim().length > 10).join('。') + '。';
  return summary.length > 200 ? summary.substring(0, 200) + '...' : summary;
}

function buildDoubleLinks(topics) {
  return topics.map(t => `[[${t}]]`).join(' ');
}

async function main() {
  console.log('=== IMA Raw → Wiki 增量编译流程 ===');
  console.log('日期:', new Date().toISOString().split('T')[0]);
  console.log('');

  console.log('步骤1: 获取raw知识库内容...');
  const rawItems = await getWikiKnowledgeList(RAW_KB_ID);
  console.log(`Raw条目总数: ${rawItems.length}`);

  const processedItems = [];
  for (const item of rawItems) {
    const title = item.title || '无标题';
    let content = item.content || '';

    if (item.url && (item.url.includes('mp.weixin.qq.com') || item.url.includes('weixin'))) {
      console.log(`  抓取微信文章: ${title}`);
      const fetched = await fetchUrlContent(item.url);
      if (fetched) {
        content = fetched.substring(0, 5000);
      }
    }

    processedItems.push({ title, content, url: item.url });
  }

  console.log(`\n步骤2-4: 分析内容、识别主题、编译wiki文件...`);
  
  const topicContents = {};
  TOPICS.forEach(t => topicContents[t] = []);
  topicContents['其他'] = [];

  for (const item of processedItems) {
    const topic = classifyTopic(item.title, item.content);
    topicContents[topic].push(item);
  }

  const wikiFiles = [];
  for (const [topic, items] of Object.entries(topicContents)) {
    if (items.length === 0) continue;
    
    let mdContent = `# ${topic}\n\n`;
    
    for (const item of items) {
      const summary = generateSummary(item.content);
      mdContent += `## ${item.title}\n\n`;
      if (summary) {
        mdContent += `> ${summary}\n\n`;
      }
      if (item.url) {
        mdContent += `[原文链接](${item.url})\n\n`;
      }
    }

    const relatedTopics = TOPICS.filter(t => t !== topic && t !== '其他');
    if (relatedTopics.length > 0) {
      mdContent += `\n---\n\n## 相关主题\n\n${buildDoubleLinks(relatedTopics)}`;
    }

    wikiFiles.push({ title: topic, content: mdContent });
    console.log(`  生成主题: ${topic} (${items.length}条)`);
  }

  console.log(`\n步骤5: 智能更新策略...`);
  
  console.log('  获取wiki当前列表...');
  const wikiList = await getWikiKnowledgeList(WIKI_KB_ID);
  const existingTitles = new Set(wikiList.map(item => item.title));
  console.log(`  Wiki当前条目数: ${existingTitles.size}`);

  let addedCount = 0;
  let updatedCount = 0;
  const failedItems = [];

  for (const file of wikiFiles) {
    try {
      if (existingTitles.has(file.title)) {
        console.log(`  已存在: ${file.title} → 追加更新`);
        
        const searchResult = await request('POST', '/note/v1/search_note', {
          search_type: 0,
          query_info: { title: file.title },
          start: 0,
          end: 1
        });

        const noteId = searchResult.data?.note_list?.[0]?.note_id;
        if (noteId) {
          const today = new Date().toISOString().split('T')[0];
          const appendContent = `\n---\n更新于 ${today}\n\n${file.content}`;
          
          await request('POST', '/note/v1/append_doc', {
            note_id: noteId,
            content_format: 1,
            content: appendContent
          });
          updatedCount++;
        } else {
          throw new Error('未找到对应笔记ID');
        }
      } else {
        console.log(`  不存在: ${file.title} → 新增`);
        
        const createResult = await request('POST', '/note/v1/import_doc', {
          content_format: 1,
          content: file.content
        });

        const noteId = createResult.data?.note_id;
        if (noteId) {
          await request('POST', '/wiki/v1/add_knowledge', {
            media_type: 11,
            note_info: { content_id: noteId },
            title: file.title,
            knowledge_base_id: WIKI_KB_ID
          });
          addedCount++;
        } else {
          throw new Error('创建笔记失败');
        }
      }
    } catch (error) {
      console.error(`  失败: ${file.title} - ${error.message}`);
      failedItems.push({ title: file.title, error: error.message });
    }
  }

  console.log(`\n步骤6: 生成报告...`);
  console.log('====================================');
  console.log('         编译报告');
  console.log('====================================');
  console.log(`Raw条目数: ${rawItems.length}`);
  console.log(`新增: ${addedCount}个`);
  console.log(`追加更新: ${updatedCount}个`);
  
  if (failedItems.length > 0) {
    console.log(`\n失败详情:`);
    failedItems.forEach(item => {
      console.log(`  - ${item.title}: ${item.error}`);
    });
  }
  
  console.log('====================================');
}

main().catch(console.error);

File: scripts/ima_compile.py
import json
import os
from datetime import datetime

RAW_KB_ID = "2Oor28yUr2eXyEFPBL0bwQU0DsrUEWPW-u9RF9CIUTw="
WIKI_KB_ID = "TcygOiKth0s91W35XJOu4NTAatl-57Bi2g19Wl3jvFw="

EXISTING_TOPICS = [
    "Claude Code与Harness Engineering",
    "AI Agent架构",
    "DeepSeek技术",
    "Prompt Engineering与AI工作流",
    "Embedding模型选型",
    "AI科研自动化",
    "个人效能与方法论",
    "知识库构建"
]

def load_feed_data():
    data = []
    
    with open('/workspace/feed-blogs.json', 'r') as f:
        blogs = json.load(f)
        for blog in blogs['blogs']:
            data.append({
                'type': 'blog',
                'title': blog['title'],
                'source': blog['name'],
                'content': blog['content'],
                'url': blog['url'],
                'published_at': blog.get('publishedAt')
            })
    
    with open('/workspace/feed-x.json', 'r') as f:
        x_data = json.load(f)
        for user in x_data['x']:
            for tweet in user['tweets']:
                data.append({
                    'type': 'tweet',
                    'title': tweet['text'][:80],
                    'source': user['name'],
                    'content': tweet['text'],
                    'url': tweet['url'],
                    'published_at': tweet['createdAt']
                })
    
    with open('/workspace/feed-podcasts.json', 'r') as f:
        podcasts = json.load(f)
        for podcast in podcasts['podcasts']:
            data.append({
                'type': 'podcast',
                'title': podcast['title'],
                'source': podcast['name'],
                'content': podcast['transcript'],
                'url': podcast['url'],
                'published_at': podcast.get('publishedAt')
            })
    
    return data

def analyze_content(raw_data):
    analysis = {
        'total_items': len(raw_data),
        'by_type': {},
        'topics': {},
        'trends': [],
        'cross_references': []
    }
    
    for item in raw_data:
        item_type = item['type']
        if item_type not in analysis['by_type']:
            analysis['by_type'][item_type] = 0
        analysis['by_type'][item_type] += 1
    
    topic_keywords = {
        'Claude Code与Harness Engineering': ['claude', 'code', 'auto mode', 'harness', 'engineering', 'permission', 'classifier', 'agentic'],
        'AI Agent架构': ['agent', 'agents', 'multi-agent', 'orchestrator', 'framework', 'architecture', 'tool use', 'tool selection'],
        'DeepSeek技术': ['deepseek'],
        'Prompt Engineering与AI工作流': ['prompt', 'prompt injection', 'workflow', 'evaluation', 'classifier', 'safety'],
        'Embedding模型选型': ['embedding', 'embeddings', 'vector', 'model selection', 'multi-modal'],
        'AI科研自动化': ['research', 'automation', 'eval', 'evaluation', 'testing', 'QA'],
        '个人效能与方法论': ['productivity', 'efficiency', 'methodology', 'system', 'workflow', 'devops', 'deepsec'],
        '知识库构建': ['knowledge', 'knowledge base', 'memory', 'retrieval', 'vector db', 'graph']
    }
    
    for item in raw_data:
        content = item['content'].lower() + ' ' + item['title'].lower()
        matched_topics = []
        
        for topic, keywords in topic_keywords.items():
            if any(keyword.lower() in content for keyword in keywords):
                matched_topics.append(topic)
                if topic not in analysis['topics']:
                    analysis['topics'][topic] = []
                analysis['topics'][topic].append(item)
        
        if matched_topics:
            for i, t1 in enumerate(matched_topics):
                for t2 in matched_topics[i+1:]:
                    ref = (t1, t2) if t1 < t2 else (t2, t1)
                    if ref not in analysis['cross_references']:
                        analysis['cross_references'].append(ref)
    
    trend_patterns = [
        ('agent adoption', ['agent', 'agents', 'multi-agent', 'orchestrator']),
        ('auto mode safety', ['auto mode', 'classifier', 'safety', 'permission']),
        ('embedding diversity', ['embedding', 'multi-modal', 'vector']),
        ('AI in education', ['education', 'student', 'class', 'teacher']),
        ('open source AI', ['open source', 'oss', 'github'])
    ]
    
    for trend_name, keywords in trend_patterns:
        count = sum(1 for item in raw_data if any(k.lower() in (item['content'] + item['title']).lower() for k in keywords))
        if count > 0:
            analysis['trends'].append({'name': trend_name, 'count': count})
    
    return analysis

def compile_wiki_files(analysis):
    wiki_dir = '/workspace/wiki'
    os.makedirs(wiki_dir, exist_ok=True)
    
    wiki_files = {}
    
    for topic, items in analysis['topics'].items():
        filename = f"{topic.replace('/', '_').replace(' ', '_')}.md"
        filepath = os.path.join(wiki_dir, filename)
        
        content = f"# {topic}\n\n"
        content += f"> 本主题共收录 {len(items)} 条内容\n\n"
        
        for item in items:
            content += f"## {item['title']}\n\n"
            content += f"> **来源**: {item['source']} | **类型**: {item['type']}\n"
            if item.get('url'):
                content += f"> **链接**: [{item['url']}]({item['url']})\n"
            if item.get('published_at'):
                content += f"> **发布时间**: {item['published_at']}\n"
            content += "\n"
            
            if item['type'] == 'tweet':
                content += item['content'] + "\n\n"
            else:
                content += item['content'][:500] + "..." if len(item['content']) > 500 else item['content']
                content += "\n\n"
        
        for other_topic in analysis['topics']:
            if other_topic != topic:
                pair = (topic, other_topic) if topic < other_topic else (other_topic, topic)
                if pair in analysis['cross_references']:
                    content += f"[[{other_topic}]]\n"
        
        wiki_files[topic] = {'filename': filename, 'filepath': filepath, 'content': content}
    
    for topic, info in wiki_files.items():
        with open(info['filepath'], 'w', encoding='utf-8') as f:
            f.write(info['content'])
    
    return wiki_files

def simulate_wiki_update(wiki_files):
    existing_notes = {
        'Claude Code与Harness Engineering': {'note_id': 'note_1', 'exists': True},
        'AI Agent架构': {'note_id': 'note_2', 'exists': True},
        '个人效能与方法论': {'note_id': 'note_3', 'exists': True}
    }
    
    results = {
        'new': [],
        'updated': [],
        'failed': []
    }
    
    today = datetime.now().strftime('%Y-%m-%d')
    
    for topic, info in wiki_files.items():
        if topic in existing_notes:
            append_content = f"\n---\n更新于 {today}\n{info['content'][:200]}..."
            results['updated'].append({
                'topic': topic,
                'note_id': existing_notes[topic]['note_id'],
                'action': 'append_doc',
                'content': append_content
            })
        else:
            results['new'].append({
                'topic': topic,
                'action': 'create_and_add',
                'content': info['content']
            })
    
    return results

def generate_report(raw_data, analysis, update_results):
    report = f"""
# IMA Raw → Wiki 增量编译报告

## 执行时间
{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}

## 一、Raw知识库概览

- **总条目数**: {analysis['total_items']}
- **条目类型分布**:
{chr(10).join([f"  - {k}: {v}条" for k, v in analysis['by_type'].items()])}

## 二、内容分析

### 2.1 纵向溯源

本批次内容涵盖多个AI领域的最新动态，包括：
- Claude Code的auto mode安全机制
- AI Agent的工具使用和架构设计
- 自动驾驶领域的技术演进（Waymo）

### 2.2 横向对比

不同来源对相同主题的观点对比：
- **AI Agent安全性**: Anthropic强调分类器和权限控制，Vercel推出deepsec进行深度安全审查
- **AI商业化**: Swyx对比OAI和Anthropic估值，Aaron Levie讨论企业AI部署趋势

### 2.3 趋势判断

{chr(10).join([f"- **{t['name']}**: {t['count']}条相关内容" for t in analysis['trends']])}

## 三、主题识别

共识别出 {len(analysis['topics'])} 个主题：

{chr(10).join([f"- **{topic}**: {len(items)}条内容" for topic, items in analysis['topics'].items()])}

## 四、Wiki更新结果

### 4.1 新增笔记
- 数量: {len(update_results['new'])}
{chr(10).join([f"  - {item['topic']}" for item in update_results['new']])}

### 4.2 追加更新
- 数量: {len(update_results['updated'])}
{chr(10).join([f"  - {item['topic']}" for item in update_results['updated']])}

### 4.3 失败记录
- 数量: {len(update_results['failed'])}

## 五、跨主题关联

{chr(10).join([f"- [[{t1}]] ↔ [[{t2}]]" for t1, t2 in analysis['cross_references']])}

## 六、备注

⚠️ **注意**: IMA API认证失败（错误码200002），本次编译基于本地feed数据模拟执行。
请检查API Key是否过期或格式是否正确，并重试完整流程。
"""
    return report

def main():
    print("步骤1: 加载raw知识库内容...")
    raw_data = load_feed_data()
    print(f"  已加载 {len(raw_data)} 条内容")
    
    print("\n步骤2: 分析内容...")
    analysis = analyze_content(raw_data)
    print(f"  识别出 {len(analysis['topics'])} 个主题")
    print(f"  发现 {len(analysis['trends'])} 个趋势")
    print(f"  发现 {len(analysis['cross_references'])} 个跨主题关联")
    
    print("\n步骤3: 编译wiki文件...")
    wiki_files = compile_wiki_files(analysis)
    print(f"  已生成 {len(wiki_files)} 个wiki文件")
    
    print("\n步骤4: 模拟Wiki更新...")
    update_results = simulate_wiki_update(wiki_files)
    print(f"  新增: {len(update_results['new'])}")
    print(f"  追加更新: {len(update_results['updated'])}")
    
    print("\n步骤5: 生成报告...")
    report = generate_report(raw_data, analysis, update_results)
    report_path = '/workspace/report.md'
    with open(report_path, 'w', encoding='utf-8') as f:
        f.write(report)
    print(f"  报告已保存到 {report_path}")
    
    print("\n" + "="*60)
    print(report)

if __name__ == '__main__':
    main()